src/transformStateWithClones.js

```diff
@@ -7,7 +7,40 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const states = [];
+  const copyState = { ...state };
+
+  for (const action of actions) {
+    const type = action.type; // string with type property
+    const addValues = action.extraData; // object with extra data
+    const removeValues = action.keysToRemove; // array with remove keys
+
+    switch (type) {
+      case 'addProperties':
+        Object.assign(copyState, addValues);
+        states.push(Object.assign({}, copyState));
+        break;
+
+      case 'removeProperties':
+        for (const value of removeValues) {
+          delete copyState[value];
+        }
+        states.push(Object.assign({}, copyState));
+        break;
+
+      case 'clear':
+        for (const value in copyState) {
+          delete copyState[value];
+        }
+        states.push(Object.assign({}, copyState));
```

Comment:
you 3 times doing Obgect.assign new object  probably you can do it one time at the end))